File: DataStructures/CPP/queue.cpp
#include <iostream>
#include <vector>

using namespace std;

class queue
{
public:
    int size;
    int front = 0;
    int rear = 0;
    vector<int> array;

    queue(int size)
    {
        this->size = size;
        this->array = vector<int>(size);
    }

    int enqueue(int element)
    {
        if (this->rear == this->size)
        {
            cout << ("Queue Overflow\n");
            return -1;
        }

        this->array[this->rear] = element;
        this->rear++;

        return 0;
    }

    int dequeue()
    {
        if (this->front == this->rear)
        {
            cout << ("Queue Underflow\n");
            return -1;
        }

        this->front++;

        return 0;
    }

    int peek()
    {
        if (this->front == this->rear)
        {
            cout << ("Queue Underflow\n");
            return -1;
        }

        return this->array[this->front];
    }

    void display()
    {
        if (this->front == this->rear)
        {
            cout << ("Queue Underflow\n");
            return;
        }

        for (int i = this->front; i < this->rear; i++)
        {
            cout << this->array[i] << " ";
        }

        cout << endl;
    }
};

int main()
{
    queue q(5);

    q.enqueue(1);
    q.enqueue(2);
    q.enqueue(3);
    q.enqueue(4);
    q.enqueue(5);

    q.display();

    q.dequeue();
    q.dequeue();

    q.display();

    cout << q.peek() << endl;

    return 0;
}

File: DataStructures/Java/Queue.java
import java.util.ArrayList;

class Queue {
    private static final int SIZE = 10;
    private static ArrayList<Integer> queue = new ArrayList<>(SIZE);
    private static int front = 0;
    private static int rear = 0;

    private static boolean isEmpty() {
        return front == rear;
    }

    private static boolean isFull() {
        return rear == SIZE;
    }

    private static void enqueue(final int data) {
        if (isFull()) {
            System.out.println("Queue is full");
            return;
        }
        queue.add(rear++, data);
    }

    private static int dequeue() {
        if (isEmpty()) {
            System.out.println("Queue is empty");
            return -1;
        }
        return queue.get(front++);
    }

    private static int peek() {
        if (isEmpty()) {
            System.out.println("Queue is empty");
            return -1;
        }
        return queue.get(front);
    }

    private static void printQueue() {
        if (isEmpty()) {
            System.out.println("Queue is empty");
            return;
        }
        System.out.print("Queue: ");
        for (int i = front; i < rear; i++) {
            System.out.print(queue.get(i) + " ");
        }
        System.out.println();
    }

    public static void main(final String[] args) {
        printQueue();

        for (int i = 1; i <= 5; i++) {
            enqueue(i);
        }

        printQueue();

        System.out.println("Dequeue: " + dequeue());
        System.out.println("Dequeue: " + dequeue());
        printQueue();

        for (int i = 6; i <= 10; i++) {
            enqueue(i);
        }

        printQueue();

        enqueue(11);
        printQueue();

        System.out.println("Peek: " + peek());
        printQueue();

        for (int i = 0; i < 5; i++) {
            System.out.println("Dequeue: " + dequeue());
        }
        printQueue();

        for (int i = 0; i < 3; i++) {
            System.out.println("Dequeue: " + dequeue());
        }

        printQueue();
        enqueue(1);
        enqueue(2);
        printQueue();
    }
}
